### ISSUE TITLE
Question about samples

### ISSUE THREAD

**Issue by Karen0103** [twelveand0/alphadiff-dataset#4]
Hello, 
I have two question about the samples. It would be nice if you could answer it.

Fristly, if I understand correctly, the following direcroty is an example of a pair of different version binaries. The 53053 directory consists of pairs of matching functions. But I don't quit understand the meaning of float value in image name such as 1.0, 0.639182, 0.790641.
![screen shot 2018-12-04 at 3 01 24 pm](https://user-images.githubusercontent.com/16543224/49469442-8702c300-f7d5-11e8-8267-312f839e5976.png)
Secondly, As discribed in the paper Section 4.2.3. For each mini-batch, 100 positive samples are selected and 200 semi- hard negative samples are generated online. In the dataset, they are all positive samples, right? Could you explain a little more about how to generate semi-hard negative samples while training?

Please feel free to contact me at ruoyan.c@uga.edu if you have any concern. Thank you!

**Comment by twelveand0** [comment]
Hello,
I'm sorry for reply so late.

Firstly, the format of a jpeg file name is:
>**index of a matching pair in database** ## **similarity measured by a tool** ## **function name** ## **sqrt(size of sample A) * sqrt(size of sample B)** ## A|B.jpeg

For example, a positive sample (matching pair) is as following:

> (4136729##1.0##queue_delayed_work_on##2 * 2##A.jpeg, 4136729##1.0##queue_delayed_work_on##2 * 2##B.jpeg) 

> the index of the pair in database is **4136729**,
>similarity of A and B is **1.0** (meaning no change), 
>function name is **queue_delayed_work_on**, 
>A's size is **2^2** and B's size is **2^2**.  

Need to note:
1. The raw bytes of a function is stored into a jpeg file without loss; you need to pull the data from a jpeg file. In fact, the jpeg format is not necessary. We stored the raw bytes into a jpeg file just because we were the first time to make use of deep learning and at last we found the image format is useless.

2. Every jpeg file is **100x100**, which is consisted of *a function's raw bytes* and *padding bytes*. The length of the raw bytes can be known by the filename (as above note); the padding value in the jpeg file is 0x90 (corresponding to NOP instruction in x86 arch). **However, during our training, we use ZERO (0x0) as the padding value and this is easy to implement in Keras.**

3. I suggest you filter out the samples whose size is no greater than 2^2, because such samples correspond to *import function*, not user-define function, and are seen bad by us. I'm very sorry for this.

4. The file "pairs" is a directory file which can be used to accelerate the initial phase of training.

The dataset was constructed only by me and is not quite good. I'm working on constructing a more perfect dataset which will involve cross-version, cross-arch, cross-compiler (maybe) and their combinations. I plan to release it before May, 2019. (ps: If you're interested, you may contact me (l.bingchang.bc@gmail.com).)

I believe the model we use in paper is not the best one and better models need to be found.

**Comment by Karen0103** [comment]
Thank you so much for your reply and detailed explaination. These information are helpful.

Just curious, as for the format of a sample file name, Is the similarity score for pair function computed by BinDiff? 

Since I am new to Deep Learning area and still confused about **negative** samples. Could you explain a little more about how to generate semi-hard negative samples while training?

**Comment by twelveand0** [comment]
About semi-hard negative samples mining: 

In the dataset, they are all positive samples (matching pairs).

For a mini-batch, firstly, 100 positive samples are selected. Secondly, semi-hard negative samples will be generated based on the 100 positive samples and the model trained until now (at the start of 1st epoch, the model's parameters have been initialized according some policy.). More specifically, for each function *F* (i.e. a element of a positive sample) in the mini-batch, all distances between it and the other functions (except the functions with the same name) will be computed with the model trained until the moment; then, we randomly select one function *F'* satisfying the condition we defined in paper as its semi-hard negative counterpart; finally, we get a negative sample by pairing *F* and *F'*. For a positive sample (i.e. two functions), we generate a semi-hard negative function for each function of the pair and get two semi-hard negative samples. So, 100 positive samples correspond to 200 negative samples.

About implementation, I have modified the core file **training.py** of Keras to implement it. The example code snippet is as following:
```python
def train_on_batch(self, x, y, sample_weight=None, class_weight=None):
    """Runs a single gradient update on a single batch of data.

    Arguments:
        x: Numpy array of training data,
            or list of Numpy arrays if the model has multiple inputs.
            If all inputs in the model are named,
            you can also pass a dictionary
            mapping input names to Numpy arrays.
        y: Numpy array of target data,
            or list of Numpy arrays if the model has multiple outputs.
            If all outputs in the model are named,
            you can also pass a dictionary
            mapping output names to Numpy arrays.
        sample_weight: optional array of the same length as x, containing
            weights to apply to the model's loss for each sample.
            In the case of temporal data, you can pass a 2D array
            with shape (samples, sequence_length),
            to apply a different weight to every timestep of every sample.
            In this case you should make sure to specify
            sample_weight_mode="temporal" in compile().
        class_weight: optional dictionary mapping
            class indices (integers) to
            a weight (float) to apply to the model's loss for the samples
            from this class during training.
            This can be useful to tell the model to "pay more attention" to
            samples from an under-represented class.

    Returns:
        Scalar training loss
        (if the model has a single output and no metrics)
        or list of scalars (if the model has multiple outputs
        and/or metrics). The attribute `model.metrics_names` will give you
        the display labels for the scalar outputs.
    """

    #===============================new added for hard negative pairs======================#
    y = np.concatenate((y, y))
    x1, x2 = x
    '''#used for triplet net
    x = [x1, x2, x1]
    '''

    '''# used for tetrad net
    x = [x1, x2, x1, x2]
    '''

    embeddings = self.positive_embeddings_on_batch(x)
    embeddings = np.concatenate((embeddings[0], embeddings[1]))
    dist_mat = pdist(embeddings)
    dist_mat = squareform(dist_mat)
    batch_x_left = []
    batch_x_right = []
    batch_y = []

    for i in range(0, x1.shape[0]+x2.shape[0]):
        if i < x1.shape[0]:
            # FaceNet semi-hard neg
            #selected_indices = np.where(np.logical_and(dist_mat[i] - dist_mat[i][i+x1.shape[0]] < 1, dist_mat[i][i+x1.shape[0]] < dist_mat[i], y != y[i]))[0]
            # VGG hard neg
            #selected_indices = np.where(np.logical_and(dist_mat[i] - dist_mat[i][i+x1.shape[0]] < 1, y != y[i]))[0]
            # semi-hard of self-design
            selected_indices = np.where(np.logical_and(dist_mat[i] < 1, dist_mat[i][i+x1.shape[0]] < dist_mat[i], y != y[i]))[0]
            # hard of self-design
            #selected_indices = np.where(np.logical_and(dist_mat[i] < 1, y != y[i]))[0]
            '''# used for tetrad net
            # Facenet semi-hard
            selected_indices_r = np.where(np.logical_and(dist_mat[i+x1.shape[0]] - dist_mat[i][i+x1.shape[0]] < 1, dist_mat[i][i+x1.shape[0]] < dist_mat[i+x1.shape[0]], y != y[i+x1.shape[0]]))[0]
            # VGG hard neg
            #selected_indices_r = np.where(np.logical_and(dist_mat[i+x1.shape[0]] - dist_mat[i][i+x1.shape[0]] < 1, y != y[i+x1.shape[0]]))[0]
            '''
        else:   
            # FaceNet semi-hard neg
            #selected_indices = np.where(np.logical_and(dist_mat[i] - dist_mat[i][i-x1.shape[0]] < 1, dist_mat[i][i-x1.shape[0]] < dist_mat[i], y != y[i]))[0]
            # VGG hard neg
            #selected_indices = np.where(np.logical_and(dist_mat[i] - dist_mat[i][i-x1.shape[0]] < 1, y != y[i]))[0]
            # semi-hard of self-design
            selected_indices = np.where(np.logical_and(dist_mat[i] < 1, dist_mat[i][i-x1.shape[0]] < dist_mat[i], y != y[i]))[0]
            # hard of self-design
            #selected_indices = np.where(np.logical_and(dist_mat[i] < 1, y != y[i]))[0]

        '''
          used for siamese net
        '''
        if selected_indices.shape[0] > 0:
            if i < x1.shape[0]:
                batch_x_left.append(x1[i])
            else:
                batch_x_left.append(x2[i-x1.shape[0]])

            rd_index_index = np.random.randint(selected_indices.shape[0])
            neg_index = selected_indices[rd_index_index]
            if neg_index < x1.shape[0]:
                batch_x_right.append(x1[neg_index])
            else:
                batch_x_right.append(x2[neg_index-x1.shape[0]])

            batch_y += [0]
        
        if i < x1.shape[0]:
            batch_x_left.append(x1[i])
            batch_x_right.append(x2[i])
            batch_y += [1]   
        
    ''' # used for siamese net'''
    batch_x_left = np.array(batch_x_left)
    batch_x_right = np.array(batch_x_right)
    x = [batch_x_left, batch_x_right]
    y = np.array(batch_y)    
 #=======================================================================================#
    x, y, sample_weights = self._standardize_user_data(
        x,
        y,
        sample_weight=sample_weight,
        class_weight=class_weight,
        check_batch_axis=True)
    if self.uses_learning_phase and not isinstance(K.learning_phase(), int):
      ins = x + y + sample_weights + [1.]
    else:
      ins = x + y + sample_weights
    self._make_train_function()
    outputs = self.train_function(ins)
    if len(outputs) == 1:
      return outputs[0]
    return outputs

def positive_embeddings_on_batch(self, x):
    """Returns predictions for a single batch of samples.

    Arguments:
        x: Input samples, as a Numpy array.

    Returns:
        Numpy array(s) of predictions.
    """
    x = _standardize_input_data(x, self._feed_input_names,
                                self._feed_input_shapes)
    if self.uses_learning_phase and not isinstance(K.learning_phase(), int):
      ins = x + [0.]
    else:
      ins = x
    self._make_predict_embedding_function()
    outputs = self.predict_function(ins)
    if len(outputs) == 1:
      return outputs[0]
    return outputs

def _make_predict_embedding_function(self):
    if not hasattr(self, 'predict_function'):
      self.predict_function = None
    if self.predict_function is None:
      if self.uses_learning_phase and not isinstance(K.learning_phase(), int):
        inputs = self._feed_inputs + [K.learning_phase()]
      else:
        inputs = self._feed_inputs
      #========================new added=====================##
      outputs = self.layers[-1].get_input_at(0)
      #========================new added=====================##
      # Gets network outputs. Does not update weights.
      # Does update the network states.
      kwargs = getattr(self, '_function_kwargs', {})
      self.predict_function = K.function(
          inputs,
          outputs,
          updates=self.state_updates,
          name='predict_function',
          **kwargs)
```

**Comment by twelveand0** [comment]
Yes, the similarity score is computed by BinDiff.

**Comment by Karen0103** [comment]
Thank you for your elaborated answers and code sharing.
